omroneipApp/src/drvOmroneip.cpp

```diff
@@ -519,1517 +590,1382 @@ asynStatus drvOmronEIP::optimiseTags()
+    status = callParamCallbacks();
+    if (status != asynSuccess)
+    {
+      asynPrint(pasynUserSelf, ASYN_TRACE_ERROR, "%s:%s Err, while performing asyn callbacks on read poller: %s\n", driverName, functionName, threadName.c_str());
     }
+    auto endTime = std::chrono::system_clock::now();
+    timeTaken = std::chrono::duration_cast<std::chrono::nanoseconds>(endTime - startTime).count() - waitTime * 1E9;
+    if (timeTaken > 0)
+      asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s:%s Poller: %s finished processing in: %d msec\n\n", driverName, functionName, threadName.c_str(), (int)(timeTaken / 1E6));
   }
-  // Check to see if array contains a valid struct
-  if (expandedMap.find(dtypeData)!=expandedMap.end()){
-    return dtypeData;
-  }
-
-  // If we get here then we have an invalid array definition
-  asynPrint(pasynUserSelf, ASYN_TRACE_ERROR, "%s:%s Invalid array definition: %s\n", driverName, functionName, arrayDesc.c_str());
-  return "Invalid";
 }
 
-int drvOmronEIP::getBiggestDtype(structDtypeMap const& expandedMap, std::string structName)
+asynStatus drvOmronEIP::writeInt8Array(asynUser *pasynUser, epicsInt8 *value, size_t nElements)
 {
-  static const char *functionName = "getBiggestDtype";
-  int thisSize = 0;
-  int biggestSize = 0;
-  std::string arrayDtype;
-  std::vector<std::string> expandedRow = expandedMap.at(structName);
-  const std::string arrayIdentifier = "\"ARRAY[";
-  for (std::string dtype : expandedRow) {
-    // Extract the dtype from an array definition string
-    if (dtype.substr(0,7) == arrayIdentifier) { //FIX ME !!!!!!
-      dtype = findArrayDtype(expandedMap, dtype);
-      if (dtype == "Invalid")
+  const char *functionName = "writeInt8Array";
+  omronDrvUser_t *drvUser = tagMap_.at(pasynUser->reason);
+  int tagIndex = drvUser->tagIndex;
+  size_t offset = drvUser->tagOffset;
+  std::string datatype = drvUser->dataType.first;
+  size_t sliceSize = drvUser->sliceSize;
+  int status = 0;
+  size_t tagSize = plc_tag_get_size(tagIndex);
+  double timeout = pasynUser->timeout * 1000;
+  bool writeOutOfBounds = 0;
+  if (nElements > tagSize)
+  {
+    // tagSize is calculated by the library based off the initial read of the tag, the user should not try and write more data than this
+    asynPrint(pasynUserSelf, ASYN_TRACE_ERROR, "%s:%s Err, with libplctag tag index: %d. Request to write more characters than can fit into the tag! nElements>tagSize:  %ld > %ld.\n",
+              driverName, functionName, tagIndex, nElements, tagSize);
+    return asynError;
+  }
+  else if (nElements < sliceSize)
+  {
+    asynPrint(pasynUserSelf, ASYN_TRACE_WARNING, "%s:%s Warn, with libplctag tag index: %d. Request to write less values than the configured sliceSize, missing data will be written as null. nElements<sliceSize:  %ld < %ld.\n",
+              driverName, functionName, tagIndex, nElements, sliceSize);
+  }
+  if (datatype == "UDT")
+  {
+    uint8_t *pOutput = (uint8_t *)malloc(tagSize * sizeof(uint8_t));
```

Comment:
How big can tagSize realistically be? If only a few bytes or KB we would be better off declaring it on the stack. If bigger, we can malloc it on the heap but then just do that once on the class rather than on every call to this method.

Reply:
Currently the max tag size would be 1994 bytes, but with future planned changes to libplctag it could be bigger. Although probably still only a few tens of KB and only in extreme cases.

Reply:
OK, if this was a read/poll function, I would lean more towards using the heap - the fast iteration makes it worth keeping the data there even into KB.
However this is a writeInt8Array function and I wouldn't expect that to be called in a fast loop. Then I'm more concerned about code maintainability: ensuring we don't forget to free. A shared pointer would be better - and even better if it can be re-used rather than re-allocated on each call.